Atomic/sign.py

```diff
@@ -0,0 +1,107 @@
+
+        signer = self.args.sign_key
+
+        for sign_image in self.args.images:
+            remote_inspect_info = util.skopeo_inspect("docker://" + sign_image)
+            manifest = util.skopeo_inspect('docker://{}'.format(sign_image), args=['--raw'], return_json=False)
+
+            # Make sure the local image is the same as the image on the
+            # registry
+            manifest_file = tempfile.NamedTemporaryFile(mode="wb", delete=False)
+            manifest_file.write(manifest)
+            manifest_file.close()
+            manifest_hash = str(util.skopeo_manifest_digest(manifest_file.name))
+
+            if use_def_sig_path:
+                _image_path = str(remote_inspect_info['Name'])
+                image_name = os.path.basename(_image_path)
+
+                _sig_path = [signature_path, os.path.dirname(_image_path), "{}@{}".format(image_name, manifest_hash)]
+                sigstore_path = str(os.path.join(*_sig_path))
+                self.make_sig_dirs(sigstore_path)
+                fq_sig_path = os.path.join(sigstore_path, self.get_sig_name(sigstore_path))
+
+            else:
+                fq_sig_path = self.args.output
+                if not os.path.exists(os.path.dirname(fq_sig_path)):
+                    raise ValueError("The path {} does not exist.".format(
+                        os.path.dirname(fq_sig_path)))
+
+            util.skopeo_standalone_sign(sign_image, manifest_file.name,
+                                        self.get_fingerprint(signer), fq_sig_path)
+
+            os.remove(manifest_file.name)
+
+    def check_input_validity(self):
+        try:
+            for image in self.args.images:
+                self._is_image(image)
+        except AtomicError:
+            raise ValueError("{} is not a valid image".format(image))
+
+    @staticmethod
+    def get_fingerprint(signer):
+        cmd = ['gpg', '--no-permission-warning', '--with-colons', '--fingerprint', signer]
+        return_code, stdout, stderr = util.subp(cmd, newline=True)
+        if return_code is not 0:
+            raise ValueError(stderr)
+        for line in stdout.splitlines():
+            if line.startswith('fpr:'):
+                return line.split(":")[9]
+
+    @staticmethod
+    def make_sig_dirs(sig_path):
+        if not os.path.exists(sig_path):
+            os.makedirs(sig_path)
+
+    @staticmethod
+    def get_sig_name(sig_path):
+        def missing_ints(aoi):
+            # Returns a list of integers in range
+            start, end = 1, max(aoi) + 1
+            if start == end and start is not 1:
+                start = 1
+            _diff = sorted(set(range(start, end)).difference(aoi))
+            if len(_diff) == 0:
+                return end
+            else:
+                return min(_diff)
+
+        sigs = []
+        for sig in os.listdir(sig_path):
+            if re.match(r"signature-\b[0-9]+\b(?!\.[0-9])", sig):
```

Comment:
What is this supposed to match? This seems like an attempt to anchor the end of the string, but what is that `\.` doing in there at all?